execute/executor.go

```diff
@@ -287,17 +300,49 @@ func (v *createExecutionNodeVisitor) Visit(node plan.Node) error {
 				}
 			}
 
-			if (plan.HasSideEffect(spec) || isParallelMerge) && len(node.Successors()) == 0 {
-				name := string(node.ID())
-				r := newResult(name)
-				v.es.results[name] = r
-				v.nodes[skipYields(node)][i].AddTransformation(r)
+			// XXX: results should be generated for terminal nodes, regardless
+			// of whether or not it is also a leaf. This is why there's an
+			// identical block at the end of each loop in each case for the
+			// "if leaf" check.
+			if len(node.Successors()) == 0 {
+				resultName := getResultName(node, spec, isParallelMerge)
+				if err := v.generateResult(resultName, node, i); err != nil {
+					return err
+				}
 			}
 		}
 	}
 	return nil
 }
 
+// generateResult will attach a result to the query for the specified node.
+func (v *createExecutionNodeVisitor) generateResult(resultName string, node plan.Node, idx int) error {
+	// if the result name is already present in the result set, that's an error.
+	if _, ok := v.es.results[resultName]; ok {
+		// XXX: we produce an error like this in the planner for duplicate yield
+		// names, but since we're generating results that aren't necessarily from
+		// yields, we need a similar check here.
+		return errors.Newf(codes.Invalid, "tried to produce more than one result with the name %q", resultName)
+	}
+	r := newResult(resultName)
+	v.es.results[resultName] = r
+	v.nodes[skipYields(node)][idx].AddTransformation(r)
+	return nil
+}
+
+// getResultName will offer a "best guess" name for a given node's result.
+//
+// For nodes that have side-effects or happen to be a Parallel Merge, the result
+// will be based on the node ID. For other cases, the default yield name of
+// `_result` is used.
+func getResultName(node plan.Node, spec plan.ProcedureSpec, isParallelMerge bool) string {
+	name := plan.DefaultYieldName
+	if plan.HasSideEffect(spec) || isParallelMerge {
```

Comment:
I [counted my chickens](https://github.com/influxdata/flux/pull/4545#discussion_r826210017) too early here. I'll leave the function in place, with a simplified body and a `TODO` mentioning this point. If you think there's a nicer way to special case the parallel names, we can do away with this function. Otherwise it'll exist mostly to encapsulate whatever you need to accomplish.